Web

thomasballinger opened the pull request:
Here's the big version! I don't know that the deploy works yet, I've got very slow upload here so it failed a couple times for me. Posting so you can take a look now, but feel free to wait until I can show a working version.

Note the last commit, it's got junk that there might be better ways to fix or at least we could use `ifdef __EMSCRIPTEN__` or `ifdef ES_GLES` for.

janisozaur committed:
Enable building without threads

Similar to https://github.com/endless-sky/endless-sky/pull/5595, remove
threads with a bunch of #ifdefs.

Now runs in Safari, but jerkily.

janisozaur committed:
HTML for web build

Add basic shell file
- Themed HTML loading screen
- Cache endless-sky.data with IndexedDB
- Allow upload of save games with HTML interface
- Load script manually, no shell html file.
- Allow downloading of save files
- Clean up HTML interface
- Add progress bar text during data download.
- Persist save files to IndexedDB
- Check if WebGL2 supported before downloading data
- Check hash and length of data and store in dataversion.js
- Report errors to stdout (#22)
- Don't fail on too small resolution (#21)
- Plugin support
  - using https://github.com/EndlessSkyCommunity/endless-sky-plugins/blob/master/generated/plugins.json
  - including a cors proxy I'll host for now
- IndexedDB shim for safari
- Catch no response.body for Pale Moon

thomasballinger committed:
emscripten build

Co-authored-by: Michał Janiszewski <janisozaur@users.noreply.github.com>

janisozaur committed:
Update credits for WebAssembly port (#24)

thomasballinger committed:
Change loading screen title from Endless Sky to Endless Web.

thomasballinger committed:
Deploy script

thomasballinger committed:
Disable music in Emscripten

thomasballinger committed:
Louder errors and more links

- more links
- Update urls.
- more space in title between Endless and Web
- top bar text
- loud errors
- Account for browsers without crossOriginIsolated implemented.
- prettier atomics error for Safari

thomasballinger committed:
hacking things till it works

thomasballinger committed:
update deploy script for new location

quyykk commented on endless-sky.html:
```diff
@@ -0,0 +1,386 @@
+<html lang="en-us">
+  <head>
+    <meta charset="utf-8">
+    <title>Endless Sky</title>
```
Do we need to change the title?

quyykk commented on source/GameWindow.cpp:
```diff
@@ -233,7 +235,7 @@ bool GameWindow::Init()
 
 	ImGui_ImplSDL2_InitForOpenGL(mainWindow, context);
-	ImGui_ImplOpenGL3_Init(nullptr);
+	ImGui_ImplOpenGL3_Init("#version 100");
```
I don't think this change is needed anymore.

thomasballinger replied:
I found I still needed it to make the imgui gui show up, I stuck it in an `#ifdef __EMSCRIPTEN__` for now

quyykk replied:
That is very surprising. I'll have debug this then (if it's even possible?). The function uses `"#version 100"` if emscripten is used so yeah very surprising.

quyykk replied:
`nullptr` works for me using the ES 3.

quyykk commented on source/SystemEditor.cpp:
```diff
@@ -1090,11 +1090,13 @@ void SystemEditor::Randomize()
 	int planetCount = randPlanetCount(gen);
 	for(int i = 0; i < planetCount; ++i)
 	{
+		/*
 		constexpr int RANDOM_SPACE = 100;
 		int space = RANDOM_SPACE;
 		for(const auto &stellar : object->objects)
 			if(!stellar.isStar && stellar.parent == -1)
 				space += RANDOM_SPACE / 2;
+				*/
```
I'm assuming this code doesn't compile?

thomasballinger replied:
Yep, I can just delete the code instead, it looks like it's not doing anything. I have
```
	-Wall\
	-Werror\
```
on in the Makefile copying the SConstruct file, maybe it wasn't alerting for you on GCC?
```
source/SystemEditor.cpp:1094:7: error: variable 'space' set but not used [-Werror,-Wunused-but-set-variable]
                int space = RANDOM_SPACE;
                    ^
1 error generated.
```

quyykk replied:
I'm not using gcc and seems like clang doesn't emit a warning here. I'll fix this

quyykk commented on source/imgui_impl_opengl3.h:
```diff
@@ -45,7 +45,7 @@ IMGUI_IMPL_API void     ImGui_ImplOpenGL3_DestroyDeviceObjects();
 // you are likely to get a crash in ImGui_ImplOpenGL3_Init().
 // You can explicitly select a loader by using one of the '#define IMGUI_IMPL_OPENGL_LOADER_XXX' in imconfig.h or compiler command-line.
-#define IMGUI_IMPL_OPENGL_LOADER_GLEW
+#define IMGUI_IMPL_OPENGL_ES2
```
This can be easily changed into "auto-detect".

Delete the whole #if thingy until the last line of the file and replace it with

```
#if defined(__EMSCRIPTEN__)
#define IMGUI_IMPL_OPENGL_ES2
#else
#define IMGUI_IMPL_OPENGL_LOADER_GLEW
#endif
```

quyykk reviewed (commented):
Thank you very much! There is a dummy file too called `dummy` can you delete it too or is it needed?

thomasballinger committed:
Leftover UI polish and CR changes

thomasballinger commented:
dummy removed, some leftover UI polish done and this is up at https://play-endless-web.com/editor

Could you give me some steps that should cause a plugin to be written to? Right now if you go to https://play-endless-web.com/editor?plugins you get to choose plugins to download from the central repository and editing these might work (an in-memory filesystem is written to) but this isn't persisted past a reload and there's not way to download them, so adding a download plugin interface is probably the minimum.

quyykk commented:
> this is up at https://play-endless-web.com/editor

I don't see the menu bar for some reason? I'm on Firefox 92.0b9 on Linux if that helps.

> Could you give me some steps that should cause a plugin to be written to?

Saving (as in writing to a file) is done in only two cases: When exiting the game or when clicking on "File -> Save All". In both cases the function that does the writing is `Editor::WriteAll`.

Also could you fix the CI error please :) thank you! :)

And I got this error when deselecting a plugin:

```
pluginPath is not defined

showPlugins/</<@https://play-endless-web.com/js/plugins.js:81:40
```

quyykk commented:
I tried compiling this PR and I got some macro redefinition errors and unknown functions being used (namely the OES extension functions). ~~I don't know exactly what happened, but this patch fixes it.~~ Now I believe it is because we are compiling with WebGL 2 only which doesn't have those extensions. ~~I tried looking it up a bit but couldn't find anything:~~

```patch
diff --git a/source/imgui_impl_opengl3.cpp b/source/imgui_impl_opengl3.cpp
index 6dedd71e..99bc73ab 100644
--- a/source/imgui_impl_opengl3.cpp
+++ b/source/imgui_impl_opengl3.cpp
@@ -103,12 +103,6 @@
 // Vertex arrays are not supported on ES2/WebGL1 unless Emscripten which uses an extension
 #ifndef IMGUI_IMPL_OPENGL_ES2
 #define IMGUI_IMPL_OPENGL_USE_VERTEX_ARRAY
-#elif defined(__EMSCRIPTEN__)
-#define IMGUI_IMPL_OPENGL_USE_VERTEX_ARRAY
-#define glBindVertexArray       glBindVertexArrayOES
-#define glGenVertexArrays       glGenVertexArraysOES
-#define glDeleteVertexArrays    glDeleteVertexArraysOES
-#define GL_VERTEX_ARRAY_BINDING GL_VERTEX_ARRAY_BINDING_OES
 #endif
 
 // Desktop GL 3.2+ has glDrawElementsBaseVertex() which GL ES and WebGL don't have.
```

quyykk commented on source/imgui_impl_opengl3.h:
```diff
@@ -45,45 +45,9 @@ IMGUI_IMPL_API void     ImGui_ImplOpenGL3_DestroyDeviceObjects();
-    #define IMGUI_IMPL_OPENGL_LOADER_GLBINDING2
+// (that's exactly what we've done here, deleted the auto-detection code and replaced with this)
+#ifdef __EMSCRIPTEN__
+#define IMGUI_IMPL_OPENGL_ES2
```
Since you're compiling with WebGL 2 which uses ES 3 (I think?) we should change this to `IMGUI_IMPL_OPENGL_ES3` and pass `nullptr` to the init function. I tested it and it works for me.

thomasballinger commented:
Ah my mistake with the link! Here's one that has the toolbar: https://play-endless-web.com/editor/index.html
edit: also https://play-endless-web.com/editor/ but I need to debug my AWS stuff to figure out why I need the trailing slash

thomasballinger committed:
more CR

thomasballinger committed:
undelete code

thomasballinger committed:
Add editor GitHub link

thomasballinger committed:
Use a cache key specific to /editor

so that switching back and forth between the two won't require
a new 130MB download each time

quyykk reviewed (approved):
LGTM. Is there anything else to do?

-- apart from merging? :)

thomasballinger commented:
Seems good to merge for me! I'm working on the plugin deselection error and going to get into plugin download, but those seem like good separate PRs.

thomasballinger commented:
Hm I'm having trouble building this now, I see
```
make: *** No rule to make target `build/emcc/imgui_demo.o', needed by `lib/emcc/libendless-sky.a'.  Stop.
```
from a make dev. I can look into it tomorrow morning or maybe this afternoon.

quyykk commented:
Don't worry about it it's me being dumb about Makefiles. While I fix this, do `git reset HEAD^`.

thomasballinger commented:
Whoops saw this late, I'm eager to learn how to write better makefiles! I was able to work around this with https://github.com/quyykk/editor/pull/3 (feel free to close that, was just showing what my problem was)

quyykk commented:
I hate Makefiles.
